presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+            }
+            verify(peerGroupLookup.get(groupId, rowId) == heapNodeIndex, "Mismatch between peer group and lookup mapping");
+
+            peerGroupIndex = peerGroupBuffer.getNextPeerIndex(peerGroupIndex);
+        }
+        while (peerGroupIndex != UNKNOWN_INDEX);
+        verify(actualPeerGroupCount == peerGroupCount, "Recorded peer group count does not match actual");
+
+        if (leftChildHeapIndex != UNKNOWN_INDEX) {
+            verify(rowComparisonStrategy.compare(rowId, peerGroupBuffer.getRowId(heapNodeBuffer.getPeerGroupIndex(leftChildHeapIndex))) > 0, "Max heap invariant violated");
+        }
+        if (rightChildHeapIndex != UNKNOWN_INDEX) {
+            verify(leftChildHeapIndex != UNKNOWN_INDEX, "Left should always be inserted before right");
+            verify(rowComparisonStrategy.compare(rowId, peerGroupBuffer.getRowId(heapNodeBuffer.getPeerGroupIndex(rightChildHeapIndex))) > 0, "Max heap invariant violated");
+        }
+
+        IntegrityStats leftIntegrityStats = verifyHeapIntegrity(groupId, leftChildHeapIndex);
+        IntegrityStats rightIntegrityStats = verifyHeapIntegrity(groupId, rightChildHeapIndex);
+
+        verify(Math.abs(leftIntegrityStats.getMaxDepth() - rightIntegrityStats.getMaxDepth()) <= 1, "Heap not balanced");
+
+        return new IntegrityStats(
+                Math.max(leftIntegrityStats.getMaxDepth(), rightIntegrityStats.getMaxDepth()) + 1,
+                leftIntegrityStats.getPeerGroupCount() + rightIntegrityStats.getPeerGroupCount() + 1,
+                leftIntegrityStats.getValueCount() + rightIntegrityStats.getValueCount() + peerGroupCount);
+    }
+
+    private static class IntegrityStats
+    {
+        private final long maxDepth;
+        private final long peerGroupCount;
+        private final long valueCount;
+
+        public IntegrityStats(long maxDepth, long peerGroupCount, long valueCount)
+        {
+            this.maxDepth = maxDepth;
+            this.peerGroupCount = peerGroupCount;
+            this.valueCount = valueCount;
+        }
+
+        public long getMaxDepth()
+        {
+            return maxDepth;
+        }
+
+        public long getPeerGroupCount()
+        {
+            return peerGroupCount;
+        }
+
+        public long getValueCount()
+        {
+            return valueCount;
+        }
+    }
+
+    /**
+     * Buffer abstracting a mapping from group ID to a heap. The group ID provides the index for all operations.
+     */
+    private static class GroupIdToHeapBuffer
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(GroupIdToHeapBuffer.class).instanceSize();
+
+        /*
+         *  Memory layout:
+         *  [LONG] heapNodeIndex1,
+         *  [LONG] heapNodeIndex2,
+         *  ...
+         */
+        // Since we have a single element per group, this array is effectively indexed on group ID
+        private final LongBigArray heapIndexBuffer = new LongBigArray(UNKNOWN_INDEX);
```

Comment:
Is this a hash map? Do we need to initialize every position to `-1`. Do we expect positions to be "freed"?

Reply:
No, this is just a straight array with positions, and freeing a position just means that it can be reused again. So in that sense there is no need to initialize everything with a default guard value.